lib/core/src/restore.rs

```diff
@@ -0,0 +1,1172 @@
+                        false => PaymentState::RefundPending,
+                    },
+                    None => PaymentState::Created,
+                },
+            },
+            None => PaymentState::Created,
+        }
+    }
+}
+
+pub(crate) struct RecoveredOnchainData {
+    send: HashMap<String, RecoveredOnchainDataSend>,
+    receive: HashMap<String, RecoveredOnchainDataReceive>,
+    chain_send: HashMap<String, RecoveredOnchainDataChainSend>,
+    chain_receive: HashMap<String, RecoveredOnchainDataChainReceive>,
+}
+
+impl LiquidSdk {
+    /// For each swap, recovers data from chain services.
+    ///
+    /// The returned data include txs and the partial swap state. This is a partial state, because
+    /// certain swap states cannot be determined based on initial and onchain data (e.g. [PaymentState::TimedOut])
+    ///
+    /// The caller is expected to merge this data with any other data available, then persist the
+    /// reconstructed swap.
+    pub(crate) async fn recover_from_onchain(&self, tx_map: TxMap) -> Result<RecoveredOnchainData> {
+        // Immutable DB, as fetched from endpoint
+        let imm_db = self.get_swaps_list().await?;
+
+        // Recovered onchain data (txs) per swap
+        let recovered = self.get_onchain_data(tx_map, &imm_db).await?;
+
+        // Validation
+        // Checks if recovered data (txs, partial state) matches known data
+        for send_swap_id in imm_db.send_swap_immutable_db_by_swap_id.keys() {
+            let full_swap = self
+                .persister
+                .fetch_send_swap_by_id(send_swap_id)
+                .ok()
+                .flatten();
+            let recovered_data = recovered.send.get(send_swap_id);
+
+            info!("[Restore Send] Validating {send_swap_id}");
+            match (full_swap, recovered_data) {
+                (Some(expected), Some(recovered)) => {
+                    let exp_lockup_tx_id = expected.lockup_tx_id;
+                    let rec_lockup_tx_id = recovered.lockup_tx_id.as_ref().map(|h| h.txid.to_hex());
+                    info!("lockup_tx_id: {exp_lockup_tx_id:?} / {rec_lockup_tx_id:?}");
+
+                    let exp_refund_tx_id = expected.refund_tx_id;
+                    let rec_refund_tx_id = recovered.refund_tx_id.as_ref().map(|h| h.txid.to_hex());
+                    info!("refund_tx_id: {exp_refund_tx_id:?} / {rec_refund_tx_id:?}");
+
+                    let exp_state = expected.state;
+                    let rec_state = recovered.get_partial_state();
+                    info!("state: {exp_state:?} / {rec_state:?}");
+                }
+                (Some(_), None) => error!("No recovered data for Send Swap {send_swap_id}"),
+                (None, Some(_)) => error!("No expected data for Send Swap {send_swap_id}"),
+                (None, None) => error!("Unknown Send Swap {send_swap_id}"),
+            }
+            info!("[Restore Send] Validated {send_swap_id}");
+        }
```

Comment:
What is the purpose of this validation step? If everything matches what we already have then why fetch on chain data from the first place? The idea is that we won't have these swaps in the db so we will have nothing to compare to.
Sames feedback goes for the validation step of other swap types.

Reply:
These validations were a way for me to check if the correct tx IDs and states are restored for various swap types.

I removed them: a0c62b802338a258fe48784fffa05b84e402ae3d

I will add real integration tests in a follow-up.